### PR TITLE
Correct hash function stub return type

### PR DESCRIPTION
**Pull request by Synchro**
Correct return type of the `hash` function. [The PHP docs](https://www.php.net/manual/en/function.hash.php) say that this function only returns a string, however this is not true. It can also return a boolean `false` if you pass in the name of a non-existent hash algorithm. For example:

    var_dump(@hash('foo', 'bar', true));
    bool(false)

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
This need a similar treatment hash_hmac got in https://github.com/phpstan/phpstan-src/pull/328.

**Comment by Synchro** [comment]
Ah. OK. At least it doesn't have the PHP 7.1 problem as hash_algos is old enough. I would have a go at that, but I'm a bit stuck for now as #109 prevents me working on a copy locally.

**Comment by ondrejmirtes** [comment]
I just merged #109 so let me know if master works for you :)

**Comment by ondrejmirtes** [comment]
Please rebase instead of creating these merge commits.

**Comment by Synchro** [comment]
Thanks for merging #109, it did indeed clean up my issue there.

I've had a go at addressing this issue, looks like it's not quite right – I think I've not understood what `NodeScopeResolverTest` does. The `$hashRandom` value is expected to be false as the algo doesn't exist – so why is it expecting `string|false` there?

I have rebased and force-pushed, but I'm not certain if it's made my earlier merge go away.

**Comment by ondrejmirtes** [comment]
I don't see the conflict - `false` is expected for `'random'` algorithm, so what's the problem?

**Comment by Synchro** [comment]
The build fails saying it's expecting `string|false` when I told it to expect `false`. At least I *think* that's what I told it!

File: src/Type/Php/HashFunctionsReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'hash';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if ($functionReflection->getName() === 'hash') {
```
Comment (ondrejmirtes) [inline src/Type/Php/HashFunctionsReturnTypeExtension.php:26]: This will always be true.

File: src/Type/Php/HashFunctionsReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			$defaultReturnType = new StringType();
+		} else {
+			$defaultReturnType = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
```
Comment (ondrejmirtes) [inline src/Type/Php/HashFunctionsReturnTypeExtension.php:29]: But we want this code be executed ;)

File: src/Type/Php/HashFunctionsReturnTypeExtension.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+
+		$values = TypeUtils::getConstantStrings($argType);
+		if (count($values) !== 1) {
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+		$string = $values[0];
+
+		return in_array($string->getValue(), hash_algos(), true) ? $defaultReturnType : new ConstantBooleanType(false);
```
Comment (ondrejmirtes) [inline src/Type/Php/HashFunctionsReturnTypeExtension.php:47]: This should be:

```php
in_array($string->getValue(), hash_algos(), true) ? new StringType() : new ConstantBooleanType(false)
```

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Just added three things that should make the tests pass :)

File: tests/PHPStan/Analyser/data/functions.php [hunk 4]
```diff
@@ -135,4 +135,7 @@
 $hashHmacFileRandom = hash_hmac_file('random', 'data', 'key');
 $hashHmacFileVariable = hash_hmac_file($string, 'data', 'key');
 
+$hash = hash('sha256', 'data', false);
+$hashRaw = hash('sha256', 'data', true);
+$hashRandom = hash('random', 'data', false);
```
[the remark this answers is not in the document]
Comment (Synchro) [reply tests/PHPStan/Analyser/data/functions.php:140]: Is that effectively what `doFoo()` returns?

[the remark this answers is not in the document]
Comment (Synchro) [reply tests/PHPStan/Analyser/data/functions.php:140]: I used the pattern used in other tests - cast the result of doFoo to mixed.

**Comment by Synchro** [comment]
So, still failing the same way 🤷‍♂️

**Comment by ondrejmirtes** [comment]
You haven't registered the extension in config.neon...

**Comment by ondrejmirtes** [comment]
It's not executed at all.

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by Synchro** [comment]
Yay! Thanks for helping me through this - it's a very complex project you've built here.

**Comment by ondrejmirtes** [comment]
It's a complex field :)

File: src/Type/Php/HashFunctionsReturnTypeExtension.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+
+		$values = TypeUtils::getConstantStrings($argType);
+		if (count($values) !== 1) {
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+		$string = $values[0];
+
+		return in_array($string->getValue(), hash_algos(), true) ? new StringType() : new ConstantBooleanType(false);
```
Comment (staabm) [inline src/Type/Php/HashFunctionsReturnTypeExtension.php:43]: Would be useful/possible to directly emit a phpstan error when `hash()` is used with a hash-algo not defined in `hash_algos()`?

Comment (ondrejmirtes) [reply src/Type/Php/HashFunctionsReturnTypeExtension.php:43]: You’ll get that error if you try to use false as a string.

**Pull request merged**